### PR TITLE
[FLINK-29818] Fix the unstable test HsResultPartitionTest.testAvailability

### PR DESCRIPTION
**Pull request by reswqa**
## What is the purpose of the change

*Fix the unstable test HsResultPartitionTest.testAvailability*

## Brief change log

  - *Fix an incorrect test case in NetworkBufferTest*
  - *Calculate the release number instead of the survival number to avoid failing to release enough memory for full spilling strategy.*
  - *fix the unstable test HsResultPartitionTest.testAvailability*

## Verifying this change

This change is already covered by existing tests*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

**Comment by reswqa** [comment]
@flinkbot run azure

**Comment by reswqa** [comment]
@flinkbot run azure

**Comment by pnowojski** [comment]
@xintongsong , could you take a look so that we can merge it ASAP? It's one of the tests that's failing master constantly. If not, let's revert FLINK-28889.

**Comment by xintongsong** [comment]
@pnowojski,
My bad. I thought this was an instability rather than a constant failure. Checking this right now.

**Comment by pnowojski** [comment]
I'm not sure if it's a constant failure, but I've had it twice in a row on the same PR, so it's pretty frequent.

**Review by xintongsong — COMMENTED** [review]
@reswqa,
I have some questions regarding the 2nd hotfix commit. In order to fix the master branch asap, I tried the fix without the questioned hotfix commit. Unfortunately, it doesn't work. So I ignored the test case for now, in d8d937dbabc975b9e8bd9096646c857b2342a7d4.

cc @pnowojski

**Comment by reswqa** [comment]
@xintongsong Thanks for the review.
> I tried the fix without the questioned hotfix commit. Unfortunately, it doesn't work. 

As you said, there's a bug that we are releasing number of survive buffers, so even if the third commit set the `releaseBufferRatio` to 0, the buffer that has been spilled will still be released, so the test will may still fail.

>I don't quite get the differences between calculating the survive vs. release numbers. What's the purpose?

The change lies in the algorithm for surviving buffers. The previous algorithm will make the number of actual release buffers smaller than the expected value, thus causing the release operation to not release enough space in time. As a result, subsequent buffer request will frequently reach the threshold of `onMemoryUsageChanged`, causing a lot of overhead.

For full spilling strategy, we set a ratio (0.4 by default) and we exactly hope that triggering the release will drop 40% of the memory space, but the previous algorithm does not meet this requirement. Because we have a precondition: only the buffers in the `spill` state can be released, which makes it easy to see that the total number of buffers that can be released (e.g. `buffersInOrder. size()`) is smaller than the expected number of surviving buffers, thus causing no buffer to be released.
The new version will first calculate the number of buffers to be released. If the `expectedSubpartitionReleaseNum` is not 0 and the current subpartition has buffers that can be released, then some buffers must be released.So I changed the algorithm here.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/hybrid/HsFullSpillingStrategyTest.java [hunk 1]
```diff
@@ -149,9 +147,9 @@ void testDecideActionWithGlobalInfo() {
 
         Map<Integer, List<BufferIndexAndChannel>> expectedReleaseBuffers = new HashMap<>();
         expectedReleaseBuffers.put(
-                subpartition1, new ArrayList<>(subpartitionBuffers1.subList(0, 2)));
+                subpartition1, new ArrayList<>(subpartitionBuffers1.subList(0, 3)));
         expectedReleaseBuffers.put(
-                subpartition2, new ArrayList<>(subpartitionBuffers2.subList(1, 3)));
+                subpartition2, new ArrayList<>(subpartitionBuffers2.subList(1, 4)));
```
[the remark this answers is not in the document]
Comment (reswqa) [reply flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/hybrid/HsFullSpillingStrategyTest.java:152]: > And please check createBufferIndexAndChannelsList. It's creating memory segment and network buffer for nothing.
good catch, a previous refactoring changed `BufferWithIdentity` into `BufferIndexAndChannel`, so there is no need to create a real buffer. Unfortunately, I forgot to remove the useless code here.

**Comment by MartijnVisser** [comment]
@reswqa @xintongsong What's the status of this PR, given that it's linked a Blocker Jira?

**Comment by reswqa** [comment]
@MartijnVisser This PR is still in review. But I think it is no longer a blocker as this test has been temporarily disabled. We can keep this Jira open, but lower its level, and close it after the PR is merged.

**Review by xintongsong — APPROVED** [review]
Sorry for the late response. LGTM. Merging.

**Pull request closed without merging**